cpp/src/bitmask/null_mask.cu

```diff
@@ -128,8 +145,92 @@ CUDF_KERNEL void set_null_mask_kernel(bitmask_type* __restrict__ destination,
+                                      bool valid,
+                                      size_type number_of_mask_words)
+{
+  set_null_mask_impl(destination, begin_bit, end_bit, valid, number_of_mask_words, cg::this_grid());
+}
 }  // namespace
 
+// Set pre-allocated null masks of given bit ranges [begin_bit, end_bit) to valids, if valid==true,
+// or null, otherwise;
+void set_null_masks_bulk(cudf::host_span<bitmask_type*> bitmasks,
+                         cudf::host_span<size_type const> begin_bits,
+                         cudf::host_span<size_type const> end_bits,
+                         cudf::host_span<bool const> valids,
+                         rmm::cuda_stream_view stream)
+{
+  CUDF_FUNC_RANGE();
+
+  auto const num_bitmasks = bitmasks.size();
+
+  CUDF_EXPECTS(num_bitmasks == begin_bits.size(),
+               "Number of bitmasks and begin bits must be equal.");
+  CUDF_EXPECTS(num_bitmasks == end_bits.size(), "Number of bitmasks and end bits must be equal.");
+  CUDF_EXPECTS(num_bitmasks == valids.size(), "Number of bitmasks and valids must be equal.");
+
+  size_t average_nullmask_words = 0;
+  auto h_number_of_mask_words   = cudf::detail::make_host_vector<size_type>(num_bitmasks, stream);
+  thrust::tabulate(
+    thrust::host, h_number_of_mask_words.begin(), h_number_of_mask_words.end(), [&](auto i) {
+      CUDF_EXPECTS(begin_bits[i] >= 0, "Invalid range.");
+      CUDF_EXPECTS(begin_bits[i] <= end_bits[i], "Invalid bit range.");
+      // Return 0 if bitmask is empty
+      if (begin_bits[i] == end_bits[i]) { return size_t{0}; }
+      // Number of words in this bitmask
+      auto const num_words =
+        num_bitmask_words(end_bits[i]) - begin_bits[i] / detail::size_in_bits<bitmask_type>();
+      // Divide by num_bitmasks and update average here to avoid overflow
+      average_nullmask_words += cudf::util::div_rounding_up_safe<size_t>(num_words, num_bitmasks);
+      return num_words;
+    });
+
+  // Create device vectors from host spans
+  auto const mr     = rmm::mr::get_current_device_resource_ref();
+  auto destinations = cudf::detail::make_device_uvector_async<bitmask_type*>(bitmasks, stream, mr);
+  auto const d_begin_bits = cudf::detail::make_device_uvector_async(begin_bits, stream, mr);
+  auto const d_end_bits   = cudf::detail::make_device_uvector_async(end_bits, stream, mr);
+  auto const d_valids     = cudf::detail::make_device_uvector_async(valids, stream, mr);
+  auto const number_of_mask_words =
+    cudf::detail::make_device_uvector_async(h_number_of_mask_words, stream, mr);
+
+  // Compute block size using heuristic and launch kernel
+  constexpr size_t max_words_per_thread  = 64;
+  constexpr size_t min_threads_per_block = 128;
+  auto block_size =
+    std::max<size_t>(min_threads_per_block, (average_nullmask_words / max_words_per_thread));
+  // Round block size to nearest (ceil) power of 2
+  block_size = size_t{1} << (63 - __builtin_clzll(block_size));
```

Comment:
I would add a note to use `std::countl_zero` once we use C++20. Might as well avoid non-standard extensions if we can. Same for the below instances of `__builtin_clz`.

Reply:
Thanks. Will add a `TODO` comment at all occurrences.